### ISSUE TITLE
Language Server crashing

### ISSUE THREAD

**Issue by decimad** [microsoft/vscode-cpptools#2222]
Beginning with 0.17.4 the language server would "crash 5 times in the last 3 minutes" when parsing my (gcc-targetting) cmake project (through compile_commands). Setting debug level to debug level to "Debug" does not log any information, in fact the output window for the extension is not even opened. Any help in pinpointing the problem or getting some diagnostics would be very welcome.

**Comment by bobbrow** [comment]
What is your operating system?

You can try cloning our repo and building and running it.  [Instructions here](https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/Building%20the%20Extension.md).

Then set a breakpoint in Extension/src/LanguageServer/client.ts on the line containing `this.outputChannel.appendLine`.  When you press F5 to debug, you will see all messages coming in from the language server one at a time.  If the crash happens before any messages come in, then you can also try enabling the "Promise Rejects" and "Uncaught Exceptions" options in the TypeScript debugger.

**Comment by sean-mcmanus** [comment]
I assume you're on 0.17.6, right? We fixed some known crashes post-0.17.4.

**Comment by Kuxe** [comment]
I can confirm that the issue that @decimad describes is present on 0.17.6 using Windows 10. Setting the debug level to "Debug" doesn't log any information for me either. Similar to @decimad, I have a cmake project that generates compile.commands.json.

**Comment by decimad** [comment]
Yes, there were similar crashes reported, that were planned to be resolved with 0.17.5, yet they obviously were not 100% related. So I hoped for 0.17.6, only after this didn't fix the problem I felt like it would be reasonable to open another issue.

**Comment by bobbrow** [comment]
@Kuxe, @decimad, if either of you can follow my instructions to see if anything is being logged at all, that would be super helpful.  The trouble with VS Code is that they delete our logs when the extension crashes, so you don't always get to see them.  We might need to add another debug setting to persist the logs for such cases, but for now, breakpoints are the only option. (besides, maybe [this](https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/Archive/Enabling%20logging.md))

**Comment by sean-mcmanus** [comment]
I think we need a call stack of the crash, via attaching a debugger to the process before the crash occurs.

**Comment by decimad** [comment]
Okay, since I can't take all day at work to get myself used to debugging this, I'm currently getting the hang of it at home, where I can't reproduce the problem though.
The settings.json of the debugee instance seems to come from the debugging host instance ,vscide seemingly, yet the c_cpp_properties.json is taken from the debuggee open folder .vscode? It's always a bit fiddely to get the cpp tools use true intellisense with a compile_commands...

**Comment by decimad** [comment]
Hrmm, okay, I put a Breakpoint to extHostOutputService.ts:38 ... no error messages regarding the language server go through there (except "channel has been closed"
Then I put a breakpoint at client.ts:257 since that seems to create the language server process. All in all I have 4 ms intellisense language server related processes open, of which 1 goes away and is restarted again repeatedly. I attached to all of them with VS.Community (all exceptions activated)... I see the process fading from the attached processes, yet VS.Community does not catch an exception or crash or something. Every time it ends up in closed-handler client.ts:365.
Any hints on where to look further would be very appreciated.

**Comment by sean-mcmanus** [comment]
The crash is not with our typescript code, it's with the Microsoft.VSCode.CPP.Extension process. You can use the typescript code via setting a breakpoint after the languageClient.start() and then using a C++ debugger to attach to the language server process. You can use either VS Code to debug it, Visual Studio, or another debugger. If you're on Windows you just have to set your symbol path to "symbolSearchPath": "http://msdl.microsoft.com/download/symbols".

Oh, I didn't read your entire post -- looks like you are debugging it. You should kill all the dangling processes and close all your C/C++ files, then when you do Launch Extension, one process should start that you can attach to. With VS Community, the symbols should be found via "Microsoft Symbol Servers". What is the exit code of the process? You need to attach in "Native" debug mode (which should be the default). Are you suggesting that the process isn't crashing and is somehow exiting in some other way?

**Comment by decimad** [comment]
It looked to me as if the language server was exiting the "natural" way on first glance. I'll dig into it further tomorrow, including having a look at the exit code.

**Comment by decimad** [comment]
VS.Community reports that the process exited with return value 0, short after which the close handler is in the extension is run.

**Comment by decimad** [comment]
So, in order to sum it up. In my experiments the Language Server did not crash but dit exit early with code 0, which the extension takes for a crash. Any help pinpointing the reason for the early exit would be very appreciated, although it seems hard to do without the sources and my knowledge ;)

**Comment by sean-mcmanus** [comment]
If you attach and check for it to break on ALL C++ and Win32 Exceptions (the default doens't have them all checked), you should at least get one break during shutdown, because we throw a terminate_threads to exit. Do you see any abnormal call stacks when you do this?

![image](https://user-images.githubusercontent.com/19859882/42719659-6db878e0-86ce-11e8-9259-e99b74b4e54c.png)

Otherwise, can you try https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.17.7-insiders ? We fixed a lot of issues. The Insiders doesn't have symbols so you won't be able to see call stacks for exceptions. I have never seen a bug where our process would just exit early we'd get more actionable info.

**Comment by decimad** [comment]
This is the crash stack: (reminder to me to open an issue that c++ exceptions are off by default)

```
>	KernelBase.dll!_RaiseException@16()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!__CxxThrowException@8()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::name_id(bool &,unsigned int &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::inner_identifier(bool &,unsigned int &,bool &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::identifier_pre_ellipsis(struct an_identifier &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::direct_declarator(struct a_type_specification &,struct a_declarator_info &,struct an_identifier &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::declarator(struct a_type_specification &,struct a_declarator_info &,struct an_identifier &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::init_declarator(struct a_type_specification &,struct a_declarator_info &,struct an_identifier &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::statement(void)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::statement_block(bool,struct a_source_point &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::statement(void)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::statement_block(bool,struct a_source_point &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::compound_statement(bool,struct a_source_point &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::function_body_strict(struct an_identifier const &,struct a_source_point &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::function_body(struct an_identifier const &,struct a_source_point &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::init_declarator_list_or_function_definition(struct a_source_point const &,struct a_source_point &,struct a_type_specification &,struct a_declaration_info const &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::simple_declaration(struct a_type_specification &,class antlr::TokenRefCount<class antlr::Token>,struct a_declaration_info const &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::property_or_event_or_simple_declaration(struct a_type_specification &,class antlr::TokenRefCount<class antlr::Token>,struct a_declaration_info const &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::general_declaration(struct a_type_specification &,class antlr::TokenRefCount<class antlr::Token>,struct a_declaration_info const &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::general_declaration_with_error_recovery_inner(struct a_type_specification &,class antlr::TokenRefCount<class antlr::Token>,struct a_declaration_info const &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::general_declaration_with_error_recovery(struct a_type_specification &,class antlr::TokenRefCount<class antlr::Token>,struct a_declaration_info const &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::declaration(class antlr::TokenRefCount<class antlr::Token>,struct a_declaration_info const &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_cpp_generated::declarations(void)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!scout_parser_base<class scout_parser_cpp,class scout_parser_cpp_generated,struct scout_parser_cpp_generatedTokenTypes>::translation_unit(void)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!_run_scout_parser()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!_process_translation_unit()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!cfe_main()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!_edg_main()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!perform_antlr_parse()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!tag_parse(int,char const * *,char const *,enum edge::parse_lang,struct edge::snapshot_session const *,class a_scout_parser_interface *)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!parser::tag_parse(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::vector<char const *,class std::allocator<char const *> > &,class code_store::a_store_tag_client &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!parser::update_file_if_needed(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,bool,bool,union code_store::a_record_id<struct code_store::a_file_tag> &,bool)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!parser::tag_parse(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,bool,union code_store::a_record_id<struct code_store::a_file_tag> &)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!browse_engine::parse_file(char const *,union code_store::a_record_id<struct code_store::a_file_tag> const &,class std::unordered_set<union code_store::a_record_id<struct code_store::a_file_tag>,struct std::hash<union code_store::a_record_id<struct code_store::a_file_tag> >,struct std::equal_to<union code_store::a_record_id<struct code_store::a_file_tag> >,class std::allocator<union code_store::a_record_id<struct code_store::a_file_tag> > > &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::set<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct std::less<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,class std::allocator<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > > const &,class std::deque<class microsoft::vc::path_utf8,class std::allocator<class microsoft::vc::path_utf8> > &,class parser &,class std::function<void >,cla	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!browse_engine::parse_file(char const *,class std::function<void >,class std::shared_ptr<class include_paths>)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!vscode::message_handler::parse_file(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::function<void >,class std::shared_ptr<struct vscode::message_handler::optional_parse_file_args>)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!<Lambda>(void)()	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!std::_LaunchPad<class std::unique_ptr<class std::tuple<class std::function<void > >,struct std::default_delete<class std::tuple<class std::function<void > > > > >::_Run(class std::_LaunchPad<class std::unique_ptr<class std::tuple<class std::function<void > >,struct std::default_delete<class std::tuple<class std::function<void > > > > > *)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!std::_LaunchPad<class std::unique_ptr<class std::tuple<class std::function<void > >,struct std::default_delete<class std::tuple<class std::function<void > > > > >::_Go(void)	Unbekannt
 	Microsoft.VSCode.CPP.Extension.exe!thread_start<unsigned int (__stdcall*)(void *)>()	Unbekannt
 	kernel32.dll!@BaseThreadInitThunk@12()	Unbekannt
 	ntdll.dll!___RtlUserThreadStart@8()	Unbekannt
 	ntdll.dll!__RtlUserThreadStart@8()	Unbekannt
```

**Comment by decimad** [comment]
0.17.7 insiders does not fix this.

**Comment by sean-mcmanus** [comment]
This same issue was reported in https://github.com/Microsoft/vscode-cpptools/issues/627 (April 2017) and https://github.com/Microsoft/vscode-cpptools/issues/2117 , but we weren't able to find a resolution. The exception thrown at this location is actually "by design" (thrown in the source code for control flow) -- it is not an unhandled exception -- you can right click on the C++ Exceptions node and add the following 4 exceptions to the list and make them unchecked to stop breaking on these:
antlr::MismatchedTokenException
antlr::NoViableAltException
antlr::RecognitionException
[rethrow]
You can check the Output window to see if any other antlr exceptions are being thrown and add those to the list to skip as well.

What happens after doing that? Does the process still exit? What other exceptions are thrown?

**Comment by decimad** [comment]
So I'm only seeing the NoViableAltException and then finally a 
```
Ausnahme ausgelöst bei 0x75E9C54F in Microsoft.VSCode.CPP.Extension.exe: Microsoft C++-Ausnahme: [rethrow] bei Speicherort 0x00000000.
```
Then the language server exits with -1073740777 (0xc0000417)

Since the exit code changed now, I assume there were some changes for the insiders preview of the extension?

I do see that around or after the time of the rethrow at 0x0 our "custom" gcc compiler is loaded, presumably to query system includes and standard defines.

**Comment by sean-mcmanus** [comment]
The 0xc0000417 seems to mean that invalid parameters are being passed to a _s (safe) c runtime function (http://stackoverflow.com/questions/3467444/invalid-cruntime-parameter-itoa-s ) -- the Insiders build doesn't have symbols for a call stack, but we plan to ship the non-Insiders release today, so hopefully you should be able to get a call stack.

**Comment by decimad** [comment]
Okay, I'll stay tuned

**Comment by decimad** [comment]
If "gcc -E -v foo.c" is conceptually what you call if you parse "gcc ...." from the compile-commands... Such a call from the command line fails for me, if I do it by hand... So maybe it's the compiler that crashes, which takes the language server with it.
With today's update to master I could not catch any exceptions whatsoever before the the language server comes down again...

**Comment by decimad** [comment]
I'm trying a different compiler version now...

**Comment by decimad** [comment]
My theory is, the compiler version I had to use use until recently took the language server with it when being asked for the include directories as described above... A newer version of the compiler successfully reports the include dirs and at the same time the language server does not crash with it.

**Comment by sean-mcmanus** [comment]
I tried using a "compiler" that crashes, but it just gives me "Timed out attempting to communicate with process!" and then falls back to WSL's /usr/bin/gcc.

If you let us know how to obtain a compiler that repros this problem we could investigate further.

**Comment by decimad** [comment]
Okay, then we might be back to beginning... the compiler's error message on the prompt when I do it manually contains german umlauts (they actually originate from the OS). Might that be a cause, considering they map to negative numbers (char is signed) on windows and some c library's functions specify only positive numbers, e.g. ascii?
I'll dig further anyways next week, however I can't give that compiler away sadly. VS.Community did not catch any exceptions whatsoever though, so at the moment I'm still unsure how to proceed.

**Comment by michelleangela** [comment]
Closing this issue.

If you update to the latest version of the C/C++ extension and your issue still persists, please re-open the issue and reply with additional information that can help us investigate your issue.
